python/sdist/amici/de_export.py

```diff
@@ -2767,9 +2767,11 @@ def _process_heavisides(
         if heavisides:
             # only apply subs if necessary
             for heaviside_sympy, heaviside_amici in heavisides:
-                dxdt = dxdt.subs(heaviside_sympy, heaviside_amici)
+                dt_expanded = dt_expanded.subs(
```

Comment:
was the issue that `expand` led to manipulation of the trigger function which was then missed during this substitution?
 
 Not super happy that we manipulate the rhs by applying expand here as I think we discussed at some point we want to keep simplification optional. I don't see why the expand is necessary here in the first place. Looks like this was introduced in https://github.com/AMICI-dev/AMICI/pull/1352

Reply:
> was the issue that `expand` led to manipulation of the trigger function which was then missed during this substitution?

Correct.

> Not super happy that we manipulate the rhs by applying expand here as I think we discussed at some point we want to keep simplification optional. I don't see why the expand is necessary here in the first place. Looks like this was introduced in #1352

```
        # expanding the rhs will in general help to collect the same
        # heaviside function
```

So I guess, not expanding the expression may lead to more root functions to be tracked than strictly necessary in certain cases. I don't know how common this issue really is, and how much of a performance impact each trigger function has.

I am open to dropping `expand`.

Reply:
> Looks like this was introduced in #1352

Specifically in https://github.com/AMICI-dev/AMICI/pull/1352/commits/f27004fd029671d5b862378adace3de7ba248d3b
Doesn't really provide more context, though.

Reply:
> > was the issue that `expand` led to manipulation of the trigger function which was then missed during this substitution?
> 
> Correct.
> 
> > Not super happy that we manipulate the rhs by applying expand here as I think we discussed at some point we want to keep simplification optional. I don't see why the expand is necessary here in the first place. Looks like this was introduced in #1352
> 
> ```
>         # expanding the rhs will in general help to collect the same
>         # heaviside function
> ```
> 
> So I guess, not expanding the expression may lead to more root functions to be tracked than strictly necessary in certain cases. I don't know how common this issue really is, and how much of a performance impact each trigger function has.
> 
> I am open to dropping `expand`.

But we are anyways calling `sp.simplify` in `_get_unique_root`. If expand really makes a difference, we could do `sp.simplify(root_found.expand() - root.get_val())`, but I would be surprised if that's really the case. We can map multiple Heaviside functions to the same root finding function, so I would prefer keeping all symbolic manipulations in `_get_unique_root`.

Reply:
Well, I don't think the `expand` ever really improved anything in terms of reducing the number of root functions. Either it would simply have had no effect at all (i.e. the same roots would have been extracted with or without `expand`), or it would have prevented replacing the Heavisides because the to-be-substituted expressions weren't found in the original expression.